### ISSUE TITLE
NoNameShadowing false-positive in a SAM with receiver lambda

### ISSUE THREAD

**Issue by technoir42** [detekt/detekt#8069]
## Expected Behavior
No issues reported since there is no `it` in the outer scope.

## Observed Behavior
False-positive when a labmda with `it` is used inside a SAM-with-receiver lambda.

## Steps to Reproduce
1. Enable Kotlin SAM-with-receiver compiler plugin, e.g. by applying `kotlin-dsl` Gradle plugin:
```kotlin
plugins {
    `kotlin-dsl`
}
```

```kotlin
class MySettingsPlugin : Plugin<Settings> {
    override fun apply(settings: Settings) {
        settings.gradle.beforeProject { // aka this
            val f = objects.fileProperty()
            f.map { it.asFile.absolutePath } // Name shadowed: implicit lambda parameter 'it' [NoNameShadowing]
        }
    }
}
```
2. Use detekt with type resolution:
```
> Task :my-settings-plugin:detektMain FAILED
/Users/user/Projects/gradle-plugins/my-settings-plugin/src/main/kotlin/com/example/MySettingsPlugin.kt:10:19: Name shadowed: implicit lambda parameter 'it' [NoNameShadowing]
```

## Context
Gradle widely uses SAM with receiver pattern, so its quite easy to hit this false-positive when working on Gradle plugins.

## Your Environment
* Version of detekt used: 1.23.8

**Comment by technoir42** [comment]
Still relevant
